Refactor Crypto Schemes Implementation

laurent-girod opened the pull request:
Hello @ineiti ,

I made most the changes you requested to the crypto schemes, namely:

- Renamed methods in `mcl.ts` to be the same as the methods in mcl
- Separating the crypto scheme, following an opinion from Wouter, I put `IBEPrimitives` in `crypto.ts` and `CrowdNotifierPrimitives` in `crowdnotifier.ts`, instead of creating a `ibe.ts` file.
- Separating the benchmarks from the spec used for testing.
- Removed unnecessary initializations.
- Commented functions in `*.spec.ts`.
- Retrieved result of function calls in benchmarks to avoid Node to optimize them out.
- Used implicit syntax for object creation when needed.
- Preferred `const` to `let`.
- Removed the dead code.

I kept the methods `setupHA`, `genPreTrace`, etc as they are back in the paper (annex B1, at the end).

Commit:
Refactor crypto schemes

ineiti commented on v2/crowdnotifier.ts:
```diff
@@ -0,0 +1,220 @@
+        );
+
+        //id = H(H(info || nonce1) || cnt || nonce2)
+
```
```suggestion
```

ineiti commented on v2/crowdnotifier.ts:
```diff
@@ -0,0 +1,220 @@
+        try {
+            const mskh_raw = crypto_box_seal_open(pre_trace.mskhEnc, keys_ha.publicKey, keys_ha.privateKey);
+            mskh.deserialize(mskh_raw);
+        }
+        catch (e) {
```
```suggestion
        } catch (e) {
```

ineiti reviewed (changes requested):
Nice - can you please also:
- Squash the commits
- manually lint (I tried to add eslint, but it complains about my mcl-types...)
  - remove empty lines before the end of a method
  - remove two consecutive empty lines in methods
- have the same naming in v1 and v1_1 (to follow Wouter's preferences ;):
  - rename v1*/crypto* to v1*/crowdnotifier*

Tomorrow I'll have a look at the actual implementation ;) Where is the new document from Wouter?

ineiti commented on v2/crowdnotifier.ts:
```diff
@@ -0,0 +1,220 @@
+     * @return the LocationData to be printed in multiple QR codes
+     */
+    static genCode(info: string, pkha: Uint8Array): ILocationData {
+        const [mpkh, mskh] = IBEPrimitives.keyGen();
```
Please use the same names as in appendix B.1 of the white paper...

ineiti commented on v2/crowdnotifier.ts:
```diff
@@ -0,0 +1,220 @@
+        const nonce2 = randombytes_buf(NONCE_LENGTH);
+
+        // In the paper ctxtHA
+        const mskhEnc = crypto_box_seal(mskh.serialize(), pkha);
```
Why the name change here?

ineiti commented on v2/crowdnotifier.ts:
```diff
@@ -0,0 +1,220 @@
+     * @param pkha public key of the health authority
+     * @return the LocationData to be printed in multiple QR codes
+     */
+    static genCode(info: string, pkha: Uint8Array): ILocationData {
```
use the same parameters as in the white paper and in the same order.

ineiti replied:
Also I think we agreed with @wouterl that there shouldn't be any `string`s in the `crowdnotifier.ts`, but only `Uint8Array`s... So `info` should be of type `Uint8Array`.

ineiti commented on v2/crowdnotifier.ts:
```diff
@@ -0,0 +1,220 @@
+    static genCode(info: string, pkha: Uint8Array): ILocationData {
+        const [mpkh, mskh] = IBEPrimitives.keyGen();
+        const [mpkv, mskv] = IBEPrimitives.keyGen();
+        const mpk = mcl.add(mpkh, mpkv);
```
This for example is very difficult to read with the 'adjusted' names you did here. To verify it, I have to go back and forth between the white paper and your names, which makes it very difficult to verify. If the names are the same as in the whitepaper, it's a 5 second verification. With the name changes it's a 1 minute verification, and potentially wrong...

ineiti commented on v2/crowdnotifier.ts:
```diff
@@ -0,0 +1,220 @@
+     * @param aux free data
+     * @return user record to be stored
+     */
+    static scan(ent: mcl.G2, pEnt: IEntProof, info: string, cnt: number, aux: string): IEncryptedData {
```
`info` and `aux` should be of type `Uint8Array`.

ineiti commented on v2/crowdnotifier.ts:
```diff
@@ -0,0 +1,220 @@
+     * @return user record to be stored
+     */
+    static scan(ent: mcl.G2, pEnt: IEntProof, info: string, cnt: number, aux: string): IEncryptedData {
+        const aux_u = new TextEncoder().encode(aux);
```
Once `aux` is of type `Uint8Array`, then this will go away. We don't want to handle text encoding in the crypto part - that is a problem of it's own and should be handled in the higher layers...

ineiti commented on v2/crowdnotifier.ts:
```diff
@@ -0,0 +1,220 @@
+     * @param nonce2 an other nonce
+     * @returns an identifier
+     */
+    static genId(info: string, cnt: number, nonce1: Uint8Array, nonce2: Uint8Array): Uint8Array {
```
I would not add `genID` here, as it is never defined in the whitepaper. It should rather be a helper method, like I did before, and not part of `CrowdNotifierPrimitives`.

ineiti reviewed (changes requested):
Difficult to follow all the implementation as the names are not the same. Please
- use the exact same names
- if things are re-defined (like `const mpk = ent` in `Scan`), please do so in the code. It makes it _much_ easier to follow the implementation. The goal is to read the code and the white-paper side-by-side and being able to follow without having to have a pen and paper by the side to make sure all the differences are correct!

laurent-girod commented:
Thanks for your code review @ineiti !

Wouter should have send you the link to Overleaf.

I updated the code to follow the paper more closely. The metod `genTrace()` remains an exception, as it stil take in argument the key pair of the health authority instead of only the secret key, because libsodium requires to have both keys to decrypt ctxtHA instead of only the private key.

The strings were removed everywhere, now we are only using `Uint8Array`s, the text encoding will be handled in a higher level.

I moved `genId()` to `Helpers`.

I renamed the files in the v1 and v1.1 API and adapted the code.

And finally I squashed the commits.

ineiti commented:
Cool, very nice! Thanks for the changes. Now I 'just' need to add linting, formatting, change the types, and add some github-actions to make sure all is done well...
